go/vt/vtgate/resolver.go

```diff
@@ -235,51 +236,37 @@ func (res *Resolver) ExecuteBatchKeyspaceIds(ctx context.Context, query *proto.K
 // It retries query if new keyspace/shards are re-resolved after a retryable error.
 func (res *Resolver) ExecuteBatch(
 	ctx context.Context,
-	queries []tproto.BoundQuery,
-	keyspace string,
 	tabletType topo.TabletType,
+	asTransaction bool,
 	session *proto.Session,
-	mapToShards func(string) (string, []string, error),
-	notInTransaction bool,
+	buildBatchRequest func() (*scatterBatchRequest, error),
 ) (*tproto.QueryResultList, error) {
-	keyspace, shards, err := mapToShards(keyspace)
+	batchRequest, err := buildBatchRequest()
 	if err != nil {
 		return nil, err
 	}
 	for {
 		qrs, err := res.scatterConn.ExecuteBatch(
 			ctx,
-			queries,
-			keyspace,
-			shards,
+			batchRequest,
 			tabletType,
-			NewSafeSession(session),
-			notInTransaction)
+			asTransaction,
+			NewSafeSession(session))
+		// Don't retry transactional requests.
+		if asTransaction {
+			return qrs, err
+		}
 		if connErrorCode, ok := isConnError(err); ok && connErrorCode == tabletconn.ERR_RETRY {
-			resharding := false
-			newKeyspace, newShards, err := mapToShards(keyspace)
-			if err != nil {
-				return nil, err
+			newBatchRequest, buildErr := buildBatchRequest()
+			if buildErr != nil {
+				return nil, buildErr
 			}
-			// check keyspace change for vertical resharding
-			if newKeyspace != keyspace {
-				keyspace = newKeyspace
-				resharding = true
-			}
-			// check shards change for horizontal resharding
-			if !StrsEquals(newShards, shards) {
-				shards = newShards
-				resharding = true
+			// Use reflect to see if the request has changed.
+			if reflect.DeepEqual(*batchRequest, *newBatchRequest) {
+				return qrs, err
```

Comment:
Got it. So essentially error code is used here to determine the system behavior. Could we have proper documentation for that ?